lnwire/writer.go

```diff
@@ -0,0 +1,442 @@
+// to have concrete type (bytes.Buffer) rather than interface type (io.Write)
+// due to performance concern.
+func WritePkScript(buf *bytes.Buffer, s PkScript) error { // nolint: interfacer
+	// The largest script we'll accept is a p2wsh which is exactly
+	// 34 bytes long.
+	scriptLength := len(s)
+	if scriptLength > 34 {
+		return ErrPkScriptTooLong
+	}
+
+	return wire.WriteVarBytes(buf, 0, s)
+}
+
+// WriteOutPoint appends the outpoint to the provided buffer.
+func WriteOutPoint(buf *bytes.Buffer, p wire.OutPoint) error {
+	// Before we write anything to the buffer, check the Index is sane.
+	if p.Index > math.MaxUint16 {
+		return ErrOutpointIndexTooBig(p.Index)
+	}
+
+	var h [32]byte
+	copy(h[:], p.Hash[:])
+	if _, err := buf.Write(h[:]); err != nil {
+		return err
+	}
+
+	// Write the index using two bytes.
+	return WriteUint16(buf, uint16(p.Index))
+}
+
+// WriteTCPAddr appends the TCP address to the provided buffer, either a IPv4
+// or a IPv6.
+func WriteTCPAddr(buf *bytes.Buffer, addr *net.TCPAddr) error {
+	if addr == nil {
+		return ErrNilTCPAddress
+	}
+
+	// Make a slice of bytes to hold the data of descriptor and ip. At
+	// most, we need 17 bytes - 1 byte for the descriptor, 16 bytes for
+	// IPv6.
+	data := make([]byte, 0, 17)
+
+	if addr.IP.To4() != nil {
+		data = append(data, uint8(tcp4Addr))
+		data = append(data, addr.IP.To4()...)
+	} else {
+		data = append(data, uint8(tcp6Addr))
+		data = append(data, addr.IP.To16()...)
+	}
+
+	if _, err := buf.Write(data); err != nil {
+		return err
+	}
+
+	return WriteUint16(buf, uint16(addr.Port))
+}
+
+// WriteOnionAddr appends the onion address to the provided buffer.
+func WriteOnionAddr(buf *bytes.Buffer, addr *tor.OnionAddr) error {
+	if addr == nil {
+		return ErrNilOnionAddress
+	}
+
+	var (
+		suffixIndex int
+		descriptor  []byte
+	)
+
+	// Decide the suffixIndex and descriptor.
+	switch len(addr.OnionService) {
+	case tor.V2Len:
+		descriptor = []byte{byte(v2OnionAddr)}
+		suffixIndex = tor.V2Len - tor.OnionSuffixLen
+
+	case tor.V3Len:
+		descriptor = []byte{byte(v3OnionAddr)}
+		suffixIndex = tor.V3Len - tor.OnionSuffixLen
+
+	default:
+		return ErrUnknownServiceLength
+	}
+
+	// Decode the address.
+	host, err := tor.Base32Encoding.DecodeString(
+		addr.OnionService[:suffixIndex],
+	)
+	if err != nil {
+		return err
+	}
+
+	// Perform the actual write when the above checks passed.
+	if _, err := buf.Write(descriptor); err != nil {
+		return err
+	}
+	if _, err := buf.Write(host); err != nil {
+		return err
+	}
+
+	return WriteUint16(buf, uint16(addr.Port))
+}
+
+// WriteNetAddrs appends a slice of addresses to the provided buffer with the
+// length info.
+func WriteNetAddrs(buf *bytes.Buffer, addresses []net.Addr) error {
+	// First, we'll encode all the addresses into an intermediate
+	// buffer. We need to do this in order to compute the total
+	// length of the addresses.
+	buffer := make([]byte, 0, MaxMsgBody)
```

Comment:
Can the length not be calculated for `[]net.Addr` so a slice doesn't have to be made?

Reply:
If I understand his suggestion, we'd essentially need to run through the set of addresses twice: once to tally up how much data is needed, and a second time to write the bytes in-line. 

Doesn't this double buffering defeat the purpose of using `bytes.Buffer` here (passed in) vs making a brand new one?

Reply:
The short answer is because it won't escape to heap, running ` go build -gcflags "-m -m"` under `lnd/lnwire`,
```
./writer.go:393:16: make([]byte, 0, MaxMsgBody) does not escape
./writer.go:394:28: &bytes.Buffer{...} does not escape
```
So there're no heap allocations.

> If I understand his suggestion, we'd essentially need to run through the set of addresses twice: once to tally up how much data is needed, and a second time to write the bytes in-line.

Yeah I think we have to write the address data somewhere before we could know the length of the data. And because we need to encode length before the actual data, `[bytes for length][bytes for data]`, we need to use a temp buffer to hold it before we write it to our passed in buffer. Otherwise we could just write to the passed in buffer then calculate how much new data was written to retrieve the length info, ie, `[bytes for data][bytes for length]`.

Only if we could first write two empty bytes, then write the data, and "insert" back the length. But I don't think that's feasible with `bytes.Buffer`, as suggested [here](https://stackoverflow.com/a/40234309/5511905).

Reply:
We'd just have to iterate over the addresses and tally up the lengths based on the type encountered -- so I don't think it'd require a temporary buffer?

Another Q: If `make([]byte, 0, MaxMsgBody)` doesn't escape to the heap, then it's on the stack.  Is it a problem if multiple goroutines are calling this function then given that the stack has limited space?

Reply:
> We'd just have to iterate over the addresses and tally up the lengths based on the type encountered -- so I don't think it'd require a temporary buffer?

Yeah we could iterate twice. The first time we calculate the length and the second time we put the length and actual data into the buffer. So it depends on whether you want to trade space with time. Since it lives on the stack, I don't think we need to.

> Another Q: If make([]byte, 0, MaxMsgBody) doesn't escape to the heap, then it's on the stack. Is it a problem if multiple goroutines are calling this function then given that the stack has limited space?

Yeah either it's heap or stack I think we both have this oom problem.